diff --git a/src/surf/network_cm02.cpp b/src/surf/network_cm02.cpp
index fe4d32d..3b15bd5 100644
--- a/src/surf/network_cm02.cpp
+++ b/src/surf/network_cm02.cpp
@@ -9,6 +9,8 @@
 #include "simgrid/sg_config.h"
 
 XBT_LOG_EXTERNAL_DEFAULT_CATEGORY(surf_network);
+//chaix
+XBT_LOG_NEW_CATEGORY(chaix,"balabala");
 
 double sg_sender_gap = 0.0;
 double sg_latency_factor = 1.0; /* default value; can be set by model or from command line */
@@ -16,7 +18,7 @@ double sg_bandwidth_factor = 1.0;       /* default value; can be set by model or
 double sg_weight_S_parameter = 0.0;     /* default value; can be set by model or from command line */
 
 double sg_tcp_gamma = 0.0;
-int sg_network_crosstraffic = 0;
+int sg_network_crosstraffic = 0; 
 
 /*************
  * CallBacks *
@@ -199,6 +201,9 @@ void NetworkCm02Model::initialize()
   }
 
   m_haveGap = false;
+
+  XBT_CVERB(chaix,"src,dst,size,rate,start,latency,route");
+
 }
 
 NetworkLinkPtr NetworkCm02Model::createNetworkLink(const char *name,
@@ -366,8 +371,10 @@ ActionPtr NetworkCm02Model::communicate(RoutingEdgePtr src, RoutingEdgePtr dst,
 
   xbt_dynar_t route = xbt_dynar_new(sizeof(RoutingEdgePtr), NULL);
 
+
   XBT_IN("(%s,%s,%g,%g)", src->getName(), dst->getName(), size, rate);
 
+
   routing_platf->getRouteAndLatency(src, dst, &route, &latency);
   xbt_assert(!xbt_dynar_is_empty(route) || latency,
              "You're trying to send data from %s to %s but there is no connection at all between these two hosts.",
@@ -467,6 +474,17 @@ ActionPtr NetworkCm02Model::communicate(RoutingEdgePtr src, RoutingEdgePtr dst,
     }
   }
 
+//  chaix
+  if(XBT_LOG_ISENABLED(chaix,	xbt_log_priority_verbose )){
+	  char routeStr[200]="";
+	  xbt_dynar_foreach(route, i, _link) {
+		  link = static_cast<NetworkCm02LinkPtr>(_link);
+		  strcat(routeStr,link->getName());
+		  strcat(routeStr,"&");
+	  }
+	  XBT_CVERB(chaix,"%s,%s,%g,%g,%g,%g,%s", src->getName(), dst->getName(), size, action->m_rate,SIMIX_get_clock(),action->m_latency,routeStr);
+  }
+  
   xbt_dynar_free(&route);
   XBT_OUT();
 
